### PR TITLE
Staged_ledger, External transition defunctor

### LINKED ISSUE #2532: Defunctor Staged Ledger
Currently, it is quite difficult to make changes to Staged_ledger. This is mainly contributed to the fact that Staged_ledger has so many functor parameter inputs. To remove the functor parameter inputs, we can have the Staged_ledger use the same stubs that `Transition_frontier_test` uses.

Revisit once #2448 is closed.

Talk to @enolan before starting this. I'm doing some of this as part of #2594

### PR DESCRIPTION
**Pull request by deepthiskumar**
Deleting the following functors:

1. Staged_ledger
2. Staged_ledger_diff
3. Transaction_snark_scan_state
4. Transaction_snark_work
5. External_transition
6. Internal_transition

Now that Staged_ledger and External_transition are defunctored we can clean up a few things like type substitutions in transition frontier, bootstrap etc. I'll make another pr for that

PS: Some of the diffs look really bad even though I just deleted `Make (.....= struct`.

Checklist:

- [ ] Tests were added for the new behavior
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [X] Does this close issues? List them:

Closes #2532

### CODE REVIEW

File: src/lib/coda_transition/external_transition_intf.ml [hunk 1]
```diff
@@ -1,40 +1,43 @@
...
 open Async_kernel
-open Signature_lib
 open Coda_base
 open Coda_state
+open Signature_lib
 
-module type Internal_transition_intf = sig
-  type staged_ledger_diff
+(*module type Staged_ledger_diff_intf = sig
+  type t [@@deriving bin_io, sexp, version]
```
Comment (wu-s-john) [inline src/lib/coda_transition/external_transition_intf.ml]: remove commented code

File: src/lib/transition_frontier/transition_frontier.ml [hunk 2]
```diff
@@ -17,14 +17,13 @@ module Make (Inputs : Inputs_intf) :
...
-   and type transaction_snark_scan_state := Inputs.Staged_ledger.Scan_state.t
-   and type verifier := Inputs.Verifier.t = struct
-  open Inputs
+              External_transition.Validation.with_transition
+   and type external_transition_validated := External_transition.Validated.t
+   and type staged_ledger_diff := Staged_ledger_diff.t
+   and type staged_ledger := Staged_ledger.t
+   and type transaction_snark_scan_state := Staged_ledger.Scan_state.t
+   and type verifier := Verifier.t = struct
+  (*open Inputs*)
```
Comment (wu-s-john) [inline src/lib/transition_frontier/transition_frontier.ml]: kill comment

File: src/lib/transition_frontier_controller_tests/test_catchup_scheduler.ml [hunk 3]
```diff
@@ -62,6 +63,10 @@ let%test_module "Transition_handler.Catchup_scheduler tests" =
       let%map children = Ivar.read result_ivar in
       Rose_tree.T (root, children)
 
+    let create () =
```
Comment (wu-s-john) [inline src/lib/transition_frontier_controller_tests/test_catchup_scheduler.ml]: can this be called create_catchup_scheduler?

File: src/lib/transition_frontier_controller_tests/test_ledger_catchup.ml [hunk 4]
```diff
@@ -31,6 +32,10 @@ end)
 
 let%test_module "Ledger catchup" =
   ( module struct
+    let run () =
```
Comment (wu-s-john) [inline src/lib/transition_frontier_controller_tests/test_ledger_catchup.ml]: Likewise, for all the transition_frontier_components, can you name the create and run functions as create_<Transition_frontier_component> and run_<Transition_frontier_component>?

File: src/lib/coda_transition/external_transition_intf.ml (hunk 1 above)
Comment (deepthiskumar) [reply src/lib/coda_transition/external_transition_intf.ml]: oops

File: src/lib/best_tip_prover/best_tip_prover.ml [hunk 5]
```diff
@@ -26,10 +27,10 @@ end
 module Make (Inputs : Inputs_intf) :
   Coda_intf.Best_tip_prover_intf
   with type transition_frontier := Inputs.Transition_frontier.t
-   and type external_transition := Inputs.External_transition.t
+   and type external_transition := External_transition.t
    and type external_transition_with_initial_validation :=
-              Inputs.External_transition.with_initial_validation
-   and type verifier := Inputs.Verifier.t = struct
+              External_transition.with_initial_validation
+   and type verifier := Verifier.t = struct
```
Comment (nholland94) [inline src/lib/best_tip_prover/best_tip_prover.ml]: Now that these are all no longer part of the functor inputs, you can actually remove these abstract types entirely from the `Best_tip_prover_intf` signature. For example, for `external_transition`, you can remove the `type external_transition` from the signature and just `%s/external_transition/External_transition.t` (regexp) the rest of the signature. Then you wouldn't need to provide this type replacement everywhere we refer to the `Best_tip_prover_intf` signature.

**Comment by cmr** [comment]
Code owned stuff LGTM

File: src/lib/best_tip_prover/best_tip_prover.ml (hunk 5 above)
Comment (deepthiskumar) [reply src/lib/best_tip_prover/best_tip_prover.ml]: Oh I was meaning to do this and defunctoring other modules in a different PR (see pr description). Wasn't sure how big of a change that'd be. But I guess I could do just this one now and the defunctoring other modules after persistence lands

**Pull request merged**